src/rust/engine/src/rule_graph.rs

```diff
@@ -475,8 +451,80 @@ impl <'a> GraphMaker<'a> {
 
   fn _remove_unfulfillable_rules_and_dependents(&self,
                                                 rule_graph: RuleGraph) -> RuleGraph {
-    // TODO port this portion
-    rule_graph
+    // Removes all unfulfillable rules transitively from the roots and the dependency edges.
+    //
+    // Takes the current root rule set and dependency table and removes all rules that are not
+    // transitively fulfillable.
+    //
+    // Deforestation. Leaping from tree to tree.
+
+    // could experiment with doing this for each rule added and deduping the traversal list
+    let mut new_unfulfillable_rules = rule_graph.unfulfillable_rules.clone();
+    let mut new_dependency_edges = rule_graph.rule_dependency_edges.clone();
+    let mut new_root_dependencies = rule_graph.root_dependencies.clone();
+
+    fn update_edges_based_on_unfulfillable_entry<K>(edge_container: &mut HashMap<K, RuleEdges>,
```

Comment:
Hm! Is this an inline named function? I don't think I realized that you could do that. And looking around, it seems like you're using them in a bunch of places that I missed in the previous review.

IMO, most of them should be broken out. An inline method in a short method might be ok, but an inline method in a super long method isn't helpful.